Mathlib/NumberTheory/NumberField/Units.lean

```diff
@@ -149,4 +165,405 @@ theorem rootsOfUnity_eq_torsion [NumberField K] :
+  let v := fun w : { w : InfinitePlace K // w ≠ w₀ } => logEmbedding K (exists_unit K w).choose
+  -- To prove the result, it is enough to prove that the family `v` is linearly independent
+  suffices B.det v ≠ 0 by
+    rw [← isUnit_iff_ne_zero, ← is_basis_iff_det] at this
+    rw [← this.2]
+    exact Submodule.span_monotone (fun _ ⟨w, hw⟩ =>
+      ⟨(exists_unit K w).choose, trivial, by rw [← hw]⟩)
+  rw [Basis.det_apply]
+  -- We use a specific lemma to prove that this determinant is nonzero
+  refine det_ne_zero_of_sum_col_lt_diag (fun w => ?_)
+  simp_rw [Real.norm_eq_abs, Basis.coePiBasisFun.toMatrix_eq_transpose, Matrix.transpose_apply]
+  rw [← sub_pos, Finset.sum_congr rfl (fun x hx => abs_of_neg ?_), Finset.sum_neg_distrib,
+    sub_neg_eq_add, Finset.sum_erase_eq_sub (Finset.mem_univ _), ← add_comm_sub]
+  refine add_pos_of_nonneg_of_pos ?_ ?_
+  · rw [sub_nonneg]
+    exact le_abs_self _
+  · rw [sum_logEmbedding_component (exists_unit K w).choose]
+    refine mul_pos_of_neg_of_neg ?_ ((exists_unit K w).choose_spec _ w.prop.symm)
+    rw [mult]; split_ifs <;> norm_num
+  · refine mul_neg_of_pos_of_neg ?_ ((exists_unit K w).choose_spec x ?_)
+    · rw [mult]; split_ifs <;> norm_num
+    · exact Subtype.ext_iff_val.not.mp (Finset.ne_of_mem_erase hx)
+
+end span_top
+
+end dirichletUnitTheorem
+
+section statements
+
+variable [NumberField K]
+
+open dirichletUnitTheorem FiniteDimensional Classical
+
+/-- The unit rank of the number field `K`, it is equal to `card (InfinitePlace K) - 1`. -/
+def rank : ℕ := Fintype.card (InfinitePlace K) - 1
+
+instance instDiscrete_unitLattice : DiscreteTopology (unitLattice K) := by
+  refine discreteTopology_of_open_singleton_zero ?_
+  refine isOpen_singleton_of_finite_mem_nhds 0 (s := Metric.closedBall 0 1) ?_ ?_
+  · exact Metric.closedBall_mem_nhds _ (by norm_num)
+  · refine Set.Finite.of_finite_image ?_ (Set.injOn_of_injective Subtype.val_injective _)
+    convert unitLattice_inter_ball_finite K 1
+    ext x
+    refine ⟨?_, fun ⟨hx1, hx2⟩ => ⟨⟨x, hx1⟩, hx2, rfl⟩⟩
+    rintro ⟨x, hx, rfl⟩
+    exact ⟨Subtype.mem x, hx⟩
+
+protected theorem finrank_eq_rank :
+    finrank ℝ ({w : InfinitePlace K // w ≠ w₀} → ℝ) = Units.rank K := by
+  simp only [finrank_fintype_fun_eq_card, Fintype.card_subtype_compl,
+    Fintype.card_ofSubsingleton, rank]
+
+instance instModuleFree_unitLattice : Module.Free ℤ (unitLattice K) :=
+  Zlattice.module_free ℝ (unitLattice_span_eq_top K)
+
+instance instModuleFinite_unitLattice : Module.Finite ℤ (unitLattice K) :=
+  Zlattice.module_finite ℝ (unitLattice_span_eq_top K)
+
+@[simp]
+theorem unitLattice_rank :
+    finrank ℤ (unitLattice K) = Units.rank K := by
+  rw [← Units.finrank_eq_rank]
+  exact Zlattice.rank ℝ (unitLattice_span_eq_top K)
+
+/-- A basis of the quotient `(𝓞 K)ˣ ⧸ (torsion K)` seen as an additive ℤ-module. -/
+def basisModTorsion : Basis (Fin (rank K)) ℤ (Additive ((𝓞 K)ˣ ⧸ (torsion K))) := by
```

Comment:
Here you are just proving that the module is free of the given rank, right? I mean, the basis has nothing special (except that it exists!). We usually don't construct data in tactic mode since it's harder to control what happens.

Here it seems better to show that it is free and compute the rank (in practice what you did), and then use `chooseBasis` if you need a fixed basis.

Reply:
Yes, this is what I meant. I think it makes clearer that `basisModTorsion` is just a random basis.